File: week03-agent-loop/agent.py
"""
⭐ Week 3: 手写 Agent Loop —— 文件整理 Agent（不用任何框架）

与 Week 2 的根本区别：
  Week 2 agent.js  -> 模型"按需调用工具"（被动）
  Week 3 agent.py  -> 模型"自己探索 -> 自己规划 -> 自己执行 -> 自己反思"（主动）

四个新增能力（这是真正的 Agent 内核）：
  1. 探索环境  : list_files 先看目录里有什么
  2. 自主规划  : 根据看到的文件类型，自己决定建哪几个文件夹
  3. 执行整理  : make_dir + move_file 逐个归位
  4. 反思查漏  : 再 list_files 看一遍，确认全部归位，才调用 done 结束

运行: python agent.py "整理 sandbox/ 目录"
"""

import os, json, sys, datetime

# Windows 控制台默认 GBK，无法打印 emoji → 强制 UTF-8
if sys.platform == "win32":
    sys.stdout.reconfigure(encoding="utf-8", errors="replace")
    sys.stderr.reconfigure(encoding="utf-8", errors="replace")

# ───────────────────── 配置 ─────────────────────
BASE_URL = os.environ.get("API_BASE_URL", "https://token-plan.cn-beijing.maas.aliyuncs.com/compatible-mode/v1")
MODEL    = os.environ.get("API_MODEL", "qwen3.8-max")
API_KEY  = os.environ.get("API_KEY", "")
MAX_TURNS = 20   # 护栏：最多循环多少轮（Week 1 重点）

if not API_KEY:
    # 兜底：从 week2 的 .env 读 key（复用）
    env_path = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "week02-tool-use", ".env")
    if os.path.exists(env_path):
        for line in open(env_path, encoding="utf-8"):
            if line.strip() and not line.startswith("#") and "=" in line:
                k, v = line.strip().split("=", 1)
                os.environ[k] = v
        API_KEY = os.environ.get("API_KEY", "")

if not API_KEY:
    sys.exit("❌ 缺少 API_KEY：请设置环境变量或先完成 week02 的 .env 配置")

# 工作目录 = agent.py 所在目录下的 sandbox/
WORK_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), "sandbox")

# ───────────────────── 一、系统提示词 ─────────────────────
SYSTEM_PROMPT = f"""你是一个【文件整理 Agent】。你的工作目录是 {WORK_DIR}。

你的任务：把工作目录里的杂文件归到合适的分类文件夹里。

【重要规则】
1. 先调用 list_files 查看目录里有什么，再规划分类方案
2. 必须自己建分类文件夹（make_dir），把文件移进去（move_file）
3. 合理的分类示例：文档/图片/代码/表格/演示文稿/压缩包，但你可以根据实际文件**自由定制**
4. 完成后必须再调用 list_files 检查一遍，确认没有遗漏
5. 全部整理妥当后，调用 done 工具汇报总结，结束任务

【工具使用规范】
- 用相对路径操作（如 "图片/全家福.jpg" 或 "全家福.jpg"），不要用绝对路径
- 不要移动/删除任何以 . 开头的隐藏文件
"""

# ───────────────────── 二、工具定义 ─────────────────────
TOOLS = [
    {"type": "function", "function": {
        "name": "list_files",
        "description": "列出工作目录下的所有文件（也可以是子目录）",
        "parameters": {"type": "object",
                       "properties": {"path": {"type": "string", "description": "子目录路径，默认留空表示根目录"}},
                       "required": []}}},
    {"type": "function", "function": {
        "name": "make_dir",
        "description": "创建一个新目录",
        "parameters": {"type": "object",
                       "properties": {"path": {"type": "string", "description": "要创建的目录路径，如 图片"}},
                       "required": ["path"]}}},
    {"type": "function", "function": {
        "name": "move_file",
        "description": "把文件从 source 移动到 destination（两个都是相对路径）",
        "parameters": {"type": "object",
                       "properties": {"source": {"type": "string", "description": "源文件相对路径"},
                                      "destination": {"type": "string", "description": "目标路径，如 图片/全家福.jpg"}},
                       "required": ["source", "destination"]}}},
    {"type": "function", "function": {
        "name": "done",
        "description": "所有整理工作完成，汇报总结并结束任务",
        "parameters": {"type": "object",
                       "properties": {"summary": {"type": "string", "description": "整理结果总结"}},
                       "required": ["summary"]}}},
]

# ───────────────────── 三、工具实现 ─────────────────────
def tool_list_files(path=""):
    target = os.path.join(WORK_DIR, path) if path else WORK_DIR
    if not os.path.isdir(target):
        return {"error": f"目录不存在: {path or '(根)'}"}
    items = os.listdir(target)
    files, dirs = [], []
    for it in items:
        full = os.path.join(target, it)
        (dirs if os.path.isdir(full) else files).append(it)
    return {"目录": sorted(dirs), "文件": sorted(files), "共": len(items)}

def tool_make_dir(path):
    target = os.path.normpath(os.path.join(WORK_DIR, path))
    if not target.startswith(os.path.abspath(WORK_DIR)):
        return {"error": "禁止越权访问目录外部"}
    os.makedirs(target, exist_ok=True)
    return {"ok": True, "已创建": path}

def tool_move_file(source, destination):
    src = os.path.normpath(os.path.join(WORK_DIR, source))
    dst = os.path.normpath(os.path.join(WORK_DIR, destination))
    # 安全校验：只允许在工作目录内移动
    if not (src.startswith(os.path.abspath(WORK_DIR)) and dst.startswith(os.path.abspath(WORK_DIR))):
        return {"error": "禁止越权移动文件"}
    if not os.path.isfile(src):
        return {"error": f"源文件不存在: {source}"}
    os.makedirs(os.path.dirname(dst), exist_ok=True)
    os.rename(src, dst)
    return {"ok": True, "已移动": f"{source} → {destination}"}

def tool_done(summary):
    # 收工信号：返回一个特殊标记，主循环看到就退出
    return {"__DONE__": True, "总结": summary}

FUNCTION_MAP = {
    "list_files": tool_list_files,
    "make_dir":   tool_make_dir,
    "move_file":  tool_move_file,
    "done":       tool_done,
}

# ───────────────────── 四、LLM 调用 ─────────────────────
def chat(messages):
    import urllib.request, urllib.error
    payload = json.dumps({"model": MODEL, "messages": messages, "tools": TOOLS}).encode()
    req = urllib.request.Request(f"{BASE_URL}/chat/completions", data=payload,
        headers={"Content-Type": "application/json", "Authorization": f"Bearer {API_KEY}"})
    try:
        with urllib.request.urlopen(req, timeout=60) as r:
            return json.loads(r.read().decode())["choices"][0]["message"]
    except urllib.error.HTTPError as e:
        raise RuntimeError(f"API {e.code}: {e.read().decode()[:200]}")

# ───────────────────── 五、Agent 主循环 ⭐ ─────────────────────
def run_agent(user_msg):
    messages = [
        {"role": "system", "content": SYSTEM_PROMPT},
        {"role": "user", "content": user_msg},
    ]
    print(f"🤖 {user_msg}\n")

    for turn in range(1, MAX_TURNS + 1):
        print(f"—— Turn {turn}/{MAX_TURNS} ——")
        reply = chat(messages)

        # 没有工具调用 → 模型直接说了（可能忘了 done），把它当 завершение
        if not reply.get("tool_calls"):
            print(f"✅ 模型直接回复（未调 done）:\n{reply.get('content')}\n")
            return

        # 执行本轮所有工具调用
        for call in reply["tool_calls"]:
            name = call["function"]["name"]
            try:
                args = json.loads(call["function"].get("arguments") or "{}")
            except json.JSONDecodeError:
                args = {}

            print(f"🔧 {name}({json.dumps(args, ensure_ascii=False)})")
            try:
                result = FUNCTION_MAP[name](**args)
            except TypeError as e:
                result = {"error": f"参数不对: {e}"}
            print(f"   ↳ {json.dumps(result, ensure_ascii=False)[:180]}")

            # 关键：工具实际上也可能 throw，统一包成 {error}（Week 1 的错误回环）
            if isinstance(result, dict) and result.get("__DONE__"):
                print(f"\n🎉 完成! {result['总结']}")
                return

            # 把「助手调用」「工具结果」追加进历史
            messages.append(reply)
            messages.append({"role": "tool", "tool_call_id": call["id"], "content": json.dumps(result, ensure_ascii=False)})

    print("⛔ 达到最大轮数，强制停止")

# ───────────────────── 六、入口 ─────────────────────
if __name__ == "__main__":
    task = sys.argv[1] if len(sys.argv) > 1 else "整理这个目录里的所有文件，把它们按类型归类"
    # 运行前拍照（整理后对比用）
    before = set(os.listdir(WORK_DIR)) if os.path.isdir(WORK_DIR) else set()
    run_agent(task)

File: week02-tool-use/agent.js
/**
 * ⭐ Week 2 完整版：天气 + 计算器 智能助手
 *
 * 这是在 agent-start.js（单工具教学版）之上的完整版：
 *   1. 两个工具：calculator（本地计算）+ get_weather（Open-Meteo 免费 API）
 *   2. 完整 Agent Loop：while 循环，直到 LLM 不再调用工具
 *   3. 带系统提示词：赋予模型角色 + 工具使用规则
 *   4. 最大轮数硬限制（防死循环烧钱，Week 1 重点护栏）
 *
 * 运行: npm start  （或在 node agent.js <你的问题>）
 */

process.loadEnvFile?.();

const BASE_URL = process.env.API_BASE_URL;
const MODEL    = process.env.API_MODEL;
const API_KEY  = process.env.API_KEY;

if (!API_KEY || !BASE_URL) {
  console.error('❌ 缺少配置：请复制 .env.example 为 .env 并填写 API_KEY / API_BASE_URL');
  process.exit(1);
}

// ───────────────────── 1. 系统提示词 ─────────────────────
// 告诉模型：你是谁、能做什么、边界在哪（Week 1 提到的护栏之一）
const SYSTEM_PROMPT = `你是一个智能助手，能使用工具解决用户问题。
可用的工具：
- calculator：四则运算
- get_weather：查询任意城市实时天气（用英文城市名查）

使用规则：
1. 当问题需要计算时，必须用 calculator，不要自己心算
2. 当问题涉及天气时，必须用 get_weather 获取数据后再回答
3. 工具结果只信真实的返回值，不要编造数据
4. 如果工具返回错误信息，如实告诉用户出了什么问题`;

// ───────────────────── 2. 工具定义（JSON Schema）─────────────────────
const TOOLS = [
  {
    type: 'function',
    function: {
      name: 'calculator',
      description: '四则运算计算器，支持 + - * /',
      parameters: {
        type: 'object',
        properties: {
          a: { type: 'number', description: '第一个数' },
          b: { type: 'number', description: '第二个数' },
          op: { type: 'string', enum: ['+', '-', '*', '/'], description: '运算符' }
        },
        required: ['a', 'b', 'op']
      }
    }
  },
  {
    type: 'function',
    function: {
      name: 'get_weather',
      description: '查询指定城市的实时天气（温度、天气状况、风速）',
      parameters: {
        type: 'object',
        properties: {
          city: { type: 'string', description: '城市英文名，如 Beijing, Shanghai, New York' }
        },
        required: ['city']
      }
    }
  }
];

// ───────────────────── 3. 工具的真实实现 ─────────────────────
// ⚠️ 核心心法（Week 1 强调）：模型只决定「调哪个、传什么参」
//    真正执行的永远是这一份代码。

const FUNCTION_MAP = {
  calculator: ({ a, b, op }) => {
    const ops = { '+': (x, y) => x + y, '-': (x, y) => x - y, '*': (x, y) => x * y, '/': (x, y) => x / y };
    if (!ops[op]) throw new Error(`不支持的运算符: ${op}`);
    if (op === '/' && b === 0) throw new Error('除数不能为 0');
    return { expression: `${a} ${op} ${b}`, result: ops[op](a, b) };
  },

  // Open-Meteo：免费、无需 key。步骤：
  // 1) 城市名 → 经纬度（Open-Meteo 的地理编码接口）
  // 2) 经纬度 → 天气数据（forecast 接口）
  get_weather: async ({ city }) => {
    // 城市 → 经纬度
    const geoUrl = `https://geocoding-api.open-meteo.com/v1/search?name=${encodeURIComponent(city)}&count=1&language=zh`;
    const geoRes = await fetch(geoUrl);
    if (!geoRes.ok) throw new Error(`地理编码失败: ${geoRes.status}`);
    const geo = await geoRes.json();
    if (!geo.results?.length) throw new Error(`找不到城市: ${city}`);

    const { latitude, longitude, name } = geo.results[0];

    // 经纬度 → 天气（current = 当前天气，后续可加 daily=7日预报）
    const wxUrl = `https://api.open-meteo.com/v1/forecast?latitude=${latitude}&longitude=${longitude}&current=temperature_2m,weather_code,wind_speed_10m&timezone=auto`;
    const wxRes = await fetch(wxUrl);
    if (!wxRes.ok) throw new Error(`天气查询失败: ${wxRes.status}`);
    const wx = await wxRes.json();

    // weather_code 是 WMO 标准码，转成中文描述
    const codeMap = {
      0:'晴', 1:'大部晴朗', 2:'多云', 3:'阴',
      45:'雾', 48:'冻雾', 51:'毛毛雨', 61:'小雨', 63:'中雨', 65:'大雨',
      80:'阵雨', 95:'雷雨', 96:'雷雨伴冰雹'
    };
    const c = wx.current;

    return {
      city: name,
      temperature_c: c.temperature_2m,
      condition: codeMap[c.weather_code] ?? `代码${c.weather_code}`,
      wind_speed_kmh: c.wind_speed_10m,
      observed_at: c.time
    };
  }
};

// ───────────────────── 4. 一次 LLM 调用 ─────────────────────
async function chat(messages) {
  const res = await fetch(`${BASE_URL}/chat/completions`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', 'Authorization': `Bearer ${API_KEY}` },
    body: JSON.stringify({ model: MODEL, messages, tools: TOOLS })
  });
  if (!res.ok) {
    const txt = await res.text();
    throw new Error(`API ${res.status}: ${txt.slice(0, 300)}`);
  }
  const data = await res.json();
  return data.choices[0].message;
}

// ───────────────────── 5. Agent Loop（核心）─────────────────────
async function runAgent(userMsg, maxTurns = 6) {
  // 历史消息：系统提示词 + 用户问题
  const messages = [
    { role: 'system', content: SYSTEM_PROMPT },
    { role: 'user', content: userMsg }
  ];

  console.log(`🤖 收到: "${userMsg}"\n`);
  let turn = 0;

  while (turn < maxTurns) {
    turn++;
    console.log(`———— 第 ${turn} 轮 ————`);

    // STEP-A: 把全部历史发给 LLM，看它想干什么
    const reply = await chat(messages);

    // 情况1：LLM 不再要工具 → 输出最终答案，循环结束
    if (!reply.tool_calls?.length) {
      console.log(`✅ 模型认为信息足够，最终回答:\n${reply.content}\n`);
      return reply.content;
    }

    // 情况2：LLM 要调用工具 → 逐个执行
    for (const call of reply.tool_calls) {
      const { name, arguments: argsStr } = call.function;
      let args;
      try {
        args = JSON.parse(argsStr);
      } catch {
        args = {}; // 模型有时会返回坏 JSON，别让整个循环崩掉
      }

      console.log(`🔧 调用工具: ${name}(${JSON.stringify(args)})`);
      let output;
      try {
        // 真正的执行！调用本地函数（async 工具也支持）
        output = await FUNCTION_MAP[name](args);
        console.log(`   ↳ 返回: ${JSON.stringify(output).slice(0, 200)}`);
      } catch (err) {
        // ⚠️ 错误也塞回循环 —— 这就是 Week 1 讲的「纠错靠错误回环」
        output = { error: err.message };
        console.log(`   ↳ 工具报错: ${err.message}`);
      }

      // 把「助手要调工具」 + 「工具结果」都追加进历史，供下一轮 LLM 看到
      messages.push(reply);                          // role=assistant, 含 tool_calls
      messages.push({
        role: 'tool',
        tool_call_id: call.id,
        content: JSON.stringify(output)
      });
    }
  }

  console.log('⛔ 达到最大轮数，强制停止（护栏生效，防止死循环烧钱）');
  return undefined;
}

// ───────────────────── 6. 入口 ─────────────────────
// 支持: node agent.js "北京今天天气"
//       node agent.js "帮我算 (8+5)*12/3"
const question = process.argv[2] ?? '查询北京今天的天气，然后用计算器算一下你看天气时的心情指数（满分100，晴+30，雨-20，多云0），告诉我结果';

runAgent(question).catch((e) => { console.error('💥', e.message); process.exit(1); });